linux/SystemdMeter.c

```diff
@@ -0,0 +1,328 @@
+/*
+htop - SystemdMeter.c
+(C) 2020 Christian Göttsche
+Released under the GNU GPLv2, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "SystemdMeter.h"
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "CRT.h"
+#include "XUtils.h"
+
+
+#define INVALID_VALUE ((unsigned int)-1)
+
+typedef void sd_bus;
+typedef void sd_bus_error;
+static int (*sym_sd_bus_open_system)(sd_bus**);
+static int (*sym_sd_bus_get_property_string)(sd_bus*, const char*, const char*, const char*, const char*, sd_bus_error*, char**);
+static int (*sym_sd_bus_get_property_trivial)(sd_bus*, const char*, const char*, const char*, const char*, sd_bus_error*, char, void*);
+static sd_bus* (*sym_sd_bus_unref)(sd_bus*);
+
+static char* systemState = NULL;
+static unsigned int nFailedUnits = INVALID_VALUE;
+static unsigned int nInstalledJobs = INVALID_VALUE;
+static unsigned int nNames = INVALID_VALUE;
+static unsigned int nJobs = INVALID_VALUE;
+static void* dlopenHandle = NULL;
+static sd_bus* bus = NULL;
+
+static void SystemdMeter_done(ATTR_UNUSED Meter* this) {
+   free(systemState);
+   systemState = NULL;
+
+   if (bus && dlopenHandle) {
+      sym_sd_bus_unref(bus);
+   }
+   bus = NULL;
+
+   if (dlopenHandle) {
+      dlclose(dlopenHandle);
+      dlopenHandle = NULL;
+   }
+}
+
+static int updateViaLib(void) {
+   if (!dlopenHandle) {
+      dlopenHandle = dlopen("libsystemd.so.0", RTLD_LAZY);
+      if (!dlopenHandle)
+         goto dlfailure;
+
+      /* Clear any errors */
+      dlerror();
+
+      *(void **)(&sym_sd_bus_open_system) = dlsym(dlopenHandle, "sd_bus_open_system");
+      if (dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_get_property_string) = dlsym(dlopenHandle, "sd_bus_get_property_string");
+      if (dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_get_property_trivial) = dlsym(dlopenHandle, "sd_bus_get_property_trivial");
+      if (dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_unref) = dlsym(dlopenHandle, "sd_bus_unref");
+      if (dlerror() != NULL)
+         goto dlfailure;
```

Comment:
`dlsym` may return `NULL` without indicating an error, thus the resolved symbols should be checked to be non-NULL.

Reply:
Isn't it the other way round? (it might return NULL on success)

https://manpages.debian.org/unstable/manpages-dev/dlsym.3.en.html
> There are several scenarios when the address of a global symbol is NULL. For example, a symbol can be placed at zero address by the linker, via a linker script or with --defsym command-line option. Undefined weak symbols also have NULL value. Finally, the symbol value may be the result of a GNU indirect function (IFUNC) resolver function that returns NULL as the resolved value. In the latter case, dlsym() also returns NULL without error. However, in the former two cases, the behavior of GNU dynamic linker is inconsistent: relocation processing succeeds and the symbol can be observed to have NULL value, but dlsym() fails and dlerror() indicates a lookup error.

Reply:
Exactly. Bad phrasing on my side. ;-)

Reply:
ahh, i missed your point at first. resolved